app/instrumentation-tests/src/org/commcare/androidTests/FormEntryTest.kt

```diff
@@ -0,0 +1,300 @@
+
+        // Confirm that we can submit the form.
+        onView(withId(R.id.nav_btn_finish))
+                .perform(click())
+
+        // Check that the form now appears in saved form and not in incomplete form.
+        onView(withText("Saved"))
+                .perform(click())
+        onView(withText("Languages"))
+                .check(matches(isDisplayed()))
+        Espresso.pressBack()
+        onView(withText(startsWith("Incomplete")))
+                .perform(click())
+        onView(withText("Languages"))
+                .check(doesNotExist())
+    }
+
+    @Test
+    fun testSaveFormMenu() {
+        InstrumentationUtility.login("user_with_no_data", "123")
+        // Create an incomplete form.
+        InstrumentationUtility.openForm(0, 0)
+        closeSoftKeyboard()
+
+        // Confirm that backing out without saving goes to form list.
+        Espresso.pressBack()
+        onView(withText(R.string.do_not_save))
+                .perform(click())
+        onView(CustomMatchers.find(
+                allOf(withText("Basic Form Tests")),
+                1
+        )).check(matches(isDisplayed()))
+        onView(withText("Languages"))
+                .perform(click())
+
+        // Make changes to the form.
+        onView(withClassName(endsWith("EditText")))
+                .perform(typeText("test"))
+        closeSoftKeyboard()
+
+        // Save the form using options menu item.
+        InstrumentationUtility.openOptionsMenu()
+        onView(withText(R.string.save_all_answers))
+                .perform(click())
+
+        // Exit form using do not save.
+        Espresso.pressBack()
+        onView(withText(R.string.do_not_save))
+                .perform(click())
+        InstrumentationUtility.gotoHome()
+
+        // Open the incomplete form and confirm that the changes exists.
+        InstrumentationUtility.openFirstIncompleteForm()
+        onView(withId(R.id.jumpBeginningButton))
+                .perform(click())
+        onView(withClassName(endsWith("EditText")))
+                .check(matches(withText("test")))
+    }
+
+    @Test
+    fun testFormEntryQuirks() {
+        InstrumentationUtility.login("user_with_no_data", "123")
+        InstrumentationUtility.openForm(0, 1)
+
+        // Trigger constraint violation(require response)
+        onView(withId(R.id.nav_btn_next))
+                .perform(click())
+        onView(withText("Sorry, this response is required!"))
+                .check(matches(isDisplayed()))
+
+        // Confirm that we can save form despite violated constraint
+        closeSoftKeyboard()
+        Espresso.pressBack()
+        onView(withText(R.string.keep_changes))
+                .perform(click())
+        onView(withText(startsWith("Incomplete")))
+                .perform(click())
+        onView(withText("Constraint"))
+                .check(matches(isDisplayed()))
+    }
+
+    @SdkSuppress(maxSdkVersion = Build.VERSION_CODES.Q)
+    @Test
+    fun testSync() {
```

Comment:
Do we have devices on BS which are on >= Q ?

Reply:
Yes, we do.